### PR TITLE
Add upgrade-puppet hook

### PR DESCRIPTION
**Pull request by ekohl**
Based on the upgrade-puppet hook in katello-installer.

### CODE REVIEW

**Comment by ekohl** [comment]
So this can be considered a preview. The helper was copied verbatim from katello-installer. I'm not sure if we need/want all of it but this allows me to quickly show a PoC. Currently it's broken but I'm interested to get some initial feedback.

**Comment by ekohl** [comment]
Verified to work on CentOS 7 now.

File: hooks/pre/31-upgrade-puppet.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
+def puppet4_available?
+  success = []
+  # TODO: OS dependent + override
```
Comment (mmoll) [inline hooks/pre/31-upgrade-puppet.rb]: what's the status here?

**Comment by mmoll** [comment]
looking good visually, but I'm unsure if I really can test drive this properly for 1.15

**Comment by ekohl** [comment]
Updated, now also tested with Debian Jessie and stripped some unused code.

Regarding inclusion in 1.15: what would give you sufficient confidence?

**Comment by mmoll** [comment]
As it is a new feature that's not interferring with the default install, I'd even merge it when you say, you tested it sucessfully under the OSes we support. We can ask people in the RC announcements to test and fix things on the way, if needed....

**Comment by mmoll** [comment]
@domcleal any opinion on this PR?

File: hooks/boot/31-upgrade-puppet.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+app_option(
+  '--upgrade-puppet',
```
Comment (domcleal) [inline hooks/boot/31-upgrade-puppet.rb:2]: This should contain the target version number (4), so as to leave room for 5 and future versions.

File: hooks/boot/01-helpers.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+class Kafo::Helpers
```
Comment (domcleal) [inline hooks/boot/01-helpers.rb:1]: It shouldn't be extending Kafo's modules from here, it should use its own namespace.

File: hooks/boot/01-helpers.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  class << self
+    def module_enabled?(kafo, name)
```
Comment (domcleal) [inline hooks/boot/01-helpers.rb:3]: Already provided in hooks: https://github.com/theforeman/kafo#hooks

File: hooks/boot/01-helpers.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+              when :debug
+                'yellow'
+              else
+                level
+              end
+
+      # \ and ' characters could cause trouble in ERB, make sure to escape them
+      escaped_message = message.gsub('\\', '\\\\\\').gsub("'", %q{\\\'})
+      say "<%= color('#{escaped_message}', :#{style}) %>"
+      Kafo::KafoConfigure.logger.send(level, message)
```
Comment (domcleal) [inline hooks/boot/01-helpers.rb:22]: Similarly, `logger` should be in the hook context

File: hooks/pre/31-upgrade-puppet.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    'puppet'        => ['autosign', 'client_package', 'codedir', 'configtimeout', 'dir', 'logdir', 'rundir', 'ssldir', 'vardir', 'server_common_modules_path', 'server_default_manifest_path', 'server_dir', 'server_envs_dir', 'server_external_nodes', 'server_jruby_gem_home', 'server_package', 'server_puppetserver_dir', 'server_puppetserver_vardir', 'server_ruby_load_paths', 'server_ssl_dir'],
+    'foreman_proxy' => ['puppet_ssl_ca', 'puppet_ssl_cert', 'puppet_ssl_key', 'puppetca_cmd', 'puppetdir', 'ssl_ca', 'ssl_cert', 'ssl_key', 'ssldir'],
+  }
+
+  unless app_value(:noop)
+    params.each do |mod, names|
+      names.each do |name|
+        p = param(mod, name)
+        fail_and_exit "Failed to find parameter '#{name}' in '#{mod}'" if p.nil?
+        p.value = nil
```
Comment (domcleal) [inline hooks/pre/31-upgrade-puppet.rb:62]: Should use `.unset_value`.

**Review by domcleal — CHANGES_REQUESTED** [review]

**Comment by domcleal** [comment]
> @domcleal any opinion on this PR?

I'm unsure if this is safe enough to use and ship with the assumptions it makes about the user's environment, so aside from the technical points above, I'm not inclined to review/merge it.

File: hooks/boot/01-helpers.rb (hunk 4 above)
Comment (ekohl) [reply hooks/boot/01-helpers.rb:3]: I tried that, but didn't work

File: hooks/pre/31-upgrade-puppet.rb (hunk 6 above)
Comment (ekohl) [reply hooks/pre/31-upgrade-puppet.rb:62]: When I tried that it didn't work. It would revert back to the value that was set before.

**Comment by mmoll** [comment]
I'll need more time dive into this, but IMHO the assumptions that are made are correct for almost any Puppet 3 RPM and DEB install?

**Comment by ekohl** [comment]
Closing this since Puppet 3 -> 4 is no longer relevant. Will revisit this when merging katello-installer into foreman-installer.

**Pull request closed without merging**